[✨feat] 스크랩 색상 수정 API 구현

jsoonworld opened the pull request:
# 📄 Work Description  
- 스크랩 색상 수정 API를 구현했습니다.  
- 기존과의 통일성을 위해 추가/삭제와 동일하게 `internshipAnnouncementId`를 path로 받도록 구성했어요.  
- 스크랩이 존재하지 않거나, 동일 색상으로 수정할 경우에 대한 예외 처리도 함께 포함되어 있어요!

---

# 💭 Thoughts

이번 작업에서는 스크랩 색상 변경 기능을 추가하면서, 기존 스크랩 추가/삭제와 API 구조를 통일시켜 일관성을 유지하고자 했어요.  
덕분에 path parameter로 `internshipAnnouncementId`를 받도록 설계했습니다!

- 색상 변경 로직(`Scrap.updateColor`)에서 **같은 색상으로 요청한 경우 예외를 던지도록 처리**했어요.
  - 이 부분은 클라이언트에서 미리 막는 게 맞을지?
  - 혹은 같은 색상이더라도 200 OK로 응답하는 게 좋을지?
  - 유빈님과 함께 응답 설계 방향을 이야기해보면 좋을 것 같아요.

- 또한, 색상 변경 시 `scrapRepository.save()`를 명시적으로 호출했는데요,
  - JPA의 **변경 감지(Dirty Checking)** 으로도 충분할지?
  - 명시적으로 save 하는 게 오히려 의도를 명확히 보여주는 건 아닐지?
  - 이 부분도 팀 기준을 정해두면 다른 작업에도 일관성을 줄 수 있을 것 같아요!

# ✅ Testing Result  
<img width="375" alt="스크린샷 2025-06-03 오후 8 01 42" src="https://github.com/user-attachments/assets/bffbaa51-2d0a-41c5-8011-640a08e52242" />


# 🗂 Related Issue  
- closed #83

jsoonworld committed:
feat(Scrap) : 스크랩 색상 변경 기능 구현 및 예외 처리 추가

- Scrap 엔티티에 색상 변경 메서드(updateColor) 구현
- 동일한 색상으로 변경 시 ScrapException 발생하도록 예외 처리
- 색상 값 확인을 위한 hexColor() 메서드 유지
- 색상 필드에 대한 setter 성격의 changeColor는 별도로 유지

jsoonworld committed:
feat(ScrapRepository) : 사용자 ID와 공고 ID로 스크랩 조회 메서드 추가

- findByInternshipAnnouncementIdAndUserId(userId, internshipAnnouncementId) 메서드 정의
- 스크랩 색상 변경 등 수정 작업을 위한 개별 스크랩 조회 용도로 사용

jsoonworld committed:
feat(ScrapErrorCode) : 색상 변경 및 스크랩 미존재 예외 코드 추가

- SCRAP_NOT_FOUND: 사용자와 공고 ID 조합으로 조회된 스크랩이 없을 때 사용
- COLOR_UNSUPPORTED: 동일 색상으로 변경 시 예외 처리 용도

jsoonworld committed:
feat(ScrapUpdateRequest) : 스크랩 색상 수정 요청 DTO 생성

- 색상 값을 문자열로 받는 ScrapUpdateRequest 추가
- PATCH /api/v1/scraps/{internshipAnnouncementId} 요청 바디용 DTO

jsoonworld committed:
feat(ScrapService) : 스크랩 색상 수정 기능 구현

- updateScrap 메서드를 통해 색상 변경 처리 로직 구현
- 동일한 색상으로 변경 시 예외 발생하도록 Scrap 도메인 로직 연동
- ScrapUpdateRequest 사용하여 color 값을 받아 처리

jsoonworld committed:
feat(ScrapController) : 스크랩 색상 수정 API 추가

- PATCH /api/v1/scraps/{internshipAnnouncementId} 엔드포인트 구현
- ScrapUpdateRequest로 색상 값 입력 받아 ScrapService에 전달
- 색상 수정 성공 시 200 OK와 함께 메시지 응답

jsoonworld committed:
test(Scrap) : 스크랩 색상 변경 기능 테스트 코드 작성

- ScrapTest에서 색상 변경 성공 및 동일 색상 변경 시 예외 발생 케이스 검증
- InternshipAnnouncement 및 User 더미 데이터 구성 포함
- COLOR_UNSUPPORTED 예외 코드 동작 확인

jsoonworld committed:
test(ScrapService) : 스크랩 색상 변경 기능 테스트 추가 및 기존 테스트 정리

- Nested 클래스로 스크랩 추가와 색상 변경 테스트를 구분하여 가독성 향상
- 스크랩 색상 변경 실패(SCRAP_NOT_FOUND) 및 성공 케이스 테스트 추가
- 기존 스크랩 테스트도 Nested(ScrapTest)로 묶어 구조 개선

jsoonworld committed:
test(ScrapController) : 스크랩 색상 변경 API 테스트 추가

- PATCH /api/v1/scraps/{internshipAnnouncementId} 엔드포인트 테스트 구현
- ScrapUpdateRequest를 JSON으로 요청하여 정상 응답 여부 검증
- scrapService mock을 통해 동작 검증

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/ScrapRepository.kt:
```diff
@@ -7,4 +7,9 @@ interface ScrapRepository : JpaRepository<Scrap, Long> {
         internshipAnnouncementId: Long,
     ): Boolean
+
+    fun findByInternshipAnnouncementIdAndUserId(
```
피드백 감사합니다! 🙏
말씀 주신 것처럼 메서드 이름과 파라미터 순서가 불일치하면 Spring Data JPA 입장에서 혼동을 줄 수 있고,
성능상의 차이는 없지만, 가독성과 컨벤션 통일 측면에서 개선하는 것이 더 좋겠다고 판단했습니다.

다만, 해당 메서드는 현재 스크랩 취소 로직에서도 함께 사용되고 있어
이번 PR에서는 우선 그대로 두고, 스크랩 취소 관련 PR에서 이름과 파라미터 순서를 함께 맞추는 방식으로 반영해보겠습니다! 💡
의견 주셔서 감사합니다 :)

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/ui/api/ScrapController.kt:
```diff
@@ -35,4 +37,25 @@ class ScrapController(
+        @PathVariable internshipAnnouncementId: Long,
+        @RequestBody scrapUpdateRequest: ScrapUpdateRequest,
+    ): ResponseEntity<ApiResponse<Unit>> {
+        val userId: Long = 1 // 임시 userId
```
피드백 감사합니다! 🙏
말씀 주신 부분처럼 userId를 하드코딩하는 것은 인증이 적용되지 않은 현재 개발 단계에서만 임시로 사용한 방식입니다.

실제 서비스 적용 시점에는 @AuthenticationPrincipal을 통해 인증된 사용자 정보를 주입받아 userId를 안전하게 처리할 계획입니다.
해당 부분은 인증 도입 이후 잊지않고 리팩토링 작업을 진행하겠습니다!💡

jsoonworld committed:
chore(ScrapService): 충돌 해결 및 트랜잭션/코드 스타일 반영

jsoonworld committed:
Merge branch 'feat/#80' into feat/#83

jsoonworld committed:
feat(ScrapController) : @AuthenticationPrincipal 적용을 위한 TODO 주석 추가

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/application/scrap/ScrapService.kt:
```diff
@@ -47,4 +48,20 @@ class ScrapService(
+
+        val color = Color.from(scrapUpdateRequest.color)
+        scrap.updateColor(color)
+
+        scrapRepository.save(scrap)
```
> 색상 변경 시 scrapRepository.save()를 명시적으로 호출했는데요,
JPA의 변경 감지(Dirty Checking) 으로도 충분할지?
명시적으로 save 하는 게 오히려 의도를 명확히 보여주는 건 아닐지?

해당 질문에 대해서 이 부분을 말씀하신 게 맞을까요?
JPA의 변경감지를 신뢰할지, save()로 명시적으로 나타낼지가 고민인 부분이시군요..!

저는 JPA가 Dirty Checking을 해 주는 기능이 있다면, JPA의 장점을 살려서 save()를 따로 명시하지 않는 게 좋을 것 같다는 입장입니다!

save()가 아무 의미 없이 명시만 해 주는 코드라면 없어도 될 것 같아요..!

jsoonworld replied:
공감합니다! 더티 채킹으로 동작되는 것을 충분히 활용하는 게 좋다고 생각해요!

다만, 혹시나 리뷰를 하는 입장에서 이러한 컨벤션 없이 완전한 처리가 되지 않았다고 생각할 수 있어서 같이 컨텍스트를 맞춰가면 좋겠다고 생각해서 말씀드렸습니다!

더팅체킹을 활용하는 방향으로 고고!

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/Scrap.kt:
```diff
@@ -36,12 +38,15 @@ class Scrap private constructor(
-
     fun hexColor(): String = color.toHexString()
 
+    fun updateColor(newColor: Color) {
+        if (this.color == newColor) {
+            throw ScrapException(ScrapErrorCode.COLOR_UNSUPPORTED)
+        }
+        this.color = newColor
+    }
```
> 색상 변경 로직(Scrap.updateColor)에서 같은 색상으로 요청한 경우 예외를 던지도록 처리했어요.


이 질문도 고민이 많이 되었던 부분인데요..

 >이 부분은 클라이언트에서 미리 막는 게 맞을지?

우선, 이러한 에러는 클라이언트에서 막았다고 할지라도 서버에서도 막아두는 게 좋을 것 같아요!
(클라이언트 개발자로서... 클라이언트를 신뢰해 주면 좋긴 하지만..) 
여러 클라이언트 파트가 있다고 할 때 모두 동일한 로직을 작성하지 않을수도 있고, 서버에서 한 번 더 막아주면 안전할 것 같습니다!

leeeyubin replied:
> 혹은 같은 색상이더라도 200 OK로 응답하는 게 좋을지?

그 다음으로 같은 색상이더라도 200을 내려줘야 하느냐는 질문에 대해 생각을 하다가 들었던 의문이 있는데요..!

`PATCH`의 작업 자체가 같은 값을 저장하려고 하면 로직상 어울리지 않기 때문에 질문을 해 주신 걸까요?

해당 고민은 서비스 특징마다 달라질 것 같다는 것이 제 생각입니다!
터닝의 경우 제가 알기로 같은 색상을 눌렀을 때 수정이 가능하게 되어 있어요! 따라서 에러를 던질 필요는 없을 것 같습니다!

jsoonworld replied:
아하, 수정인데 같은 값으로 변경을 요구하는 것이 서버 자원 낭비라고 생각되어서 말씀드렸습니다! 

그럼에도 불구하고 이 역시 변경을 하게 되면 클라이언트의 변경이 요구되고, 클라이언트분들께 변경을 부탁드리는 비용보다 현재는 서버 자원 소비가 효용이 더 클 것으로 예상되기에.. 다시 이 부분은 개선해 보겠습니다!

leeeyubin reviewed (approved):
스크랩 수정 API도 잘 확인했습니다!
남겨주신 질문 관련해서 제 생각도 남겨보았는데 확인 부탁드려요☺️

jsoonworld committed:
refactor(Scrap) : 같은 색상으로 색상 변경 시 예외 대신 무시하도록 개선

- Scrap.updateColor 메서드에서 동일한 색상으로 요청할 경우 예외를 던지지 않고 무시하도록 변경
- ScrapErrorCode.COLOR_UNSUPPORTED 항목 제거
- 관련 테스트 코드 제거 (같은 색상 변경 시 예외 발생 테스트)
